src/element/textElement.ts

```diff
@@ -1,12 +1,385 @@
+                element.width,
+              );
+              console.info("called wrap text");
+            }
+          }
+
+          const dimensions = measureText(text, getFontString(textElement));
+          nextWidth = dimensions.width;
+          nextHeight = dimensions.height;
+          nextBaseLine = dimensions.baseline;
+        }
+        // increase height in case text element height exceeds
+        if (nextHeight > element.height - PADDING * 2) {
+          containerHeight = nextHeight + PADDING * 2;
+          mutateElement(element, { height: containerHeight });
+        }
+
+        const updatedY = element!.y + containerHeight / 2 - nextHeight / 2;
+
+        const updatedX = element!.x + element!.width / 2 - nextWidth / 2;
+        mutateElement(textElement, {
+          text,
+          width: nextWidth,
+          height: nextHeight,
+          y: updatedY,
+          x: updatedX,
+          baseline: nextBaseLine,
+        });
+      }
+    }
+  });
+};
+
+// https://github.com/grassator/canvas-text-editor/blob/master/lib/FontMetrics.js
+export const measureText = (
+  text: string,
+  font: FontString,
+  maxWidth?: number | null,
+) => {
+  text = text
+    .split("\n")
+    // replace empty lines with single space because leading/trailing empty
+    // lines would be stripped from computation
+    .map((x) => x || " ")
+    .join("\n");
+  const container = document.createElement("div");
+  container.style.position = "absolute";
+  container.style.whiteSpace = "pre";
+  container.style.font = font;
+
+  if (maxWidth) {
+    const lineHeight = getApproxLineHeight(font);
+    container.style.width = `${String(maxWidth)}px`;
+    container.style.maxWidth = `${String(maxWidth)}px`;
+    container.style.overflow = "hidden";
+    container.style.wordBreak = "break-word";
+    container.style.lineHeight = `${String(lineHeight)}px`;
+    container.style.whiteSpace = "normal";
+  }
+  document.body.appendChild(container);
+  container.innerText = text;
+
+  const span = document.createElement("span");
+  span.style.display = "inline-block";
+  span.style.overflow = "hidden";
+  span.style.width = "1px";
+  span.style.height = "1px";
+  container.appendChild(span);
+  // Baseline is important for positioning text on canvas
+  const baseline = span.offsetTop + span.offsetHeight;
+  const width = container.offsetWidth;
+
+  const height = container.offsetHeight;
+  document.body.removeChild(container);
+
+  return { width, height, baseline };
+};
+
+const DUMMY_TEXT = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".toLocaleUpperCase();
+export const getApproxLineHeight = (font: FontString) => {
+  return measureText(DUMMY_TEXT, font, null).height;
+};
+
+let canvas: HTMLCanvasElement | undefined;
+const getTextWidth = (text: string, font: FontString) => {
+  if (!canvas) {
+    canvas = document.createElement("canvas");
+  }
+  const canvas2dContext = canvas.getContext("2d") as CanvasRenderingContext2D;
+  canvas2dContext.font = font;
```

Comment:
```suggestion
  const canvas2dContext = canvas.getContext("2d")!;
  canvas2dContext.font = font;
```

Reply:
I though using type is better than `!`, isn't it ?